contracts/LiquidityTree.sol

```diff
@@ -516,6 +529,91 @@ contract LiquidityTree {
+     *      Returns `true` - means found exception and need to update whole leaves
+     * @param node - start from node
+     * @param begin - node left element
+     * @param end - node right element
+     * @param l - left leaf child
+     * @param r - right leaf child
+     * @param amount - amount to add/reduce stored amounts
+     * @param isSub - true means negative to reduce
+     * @return isUpdateInsufficient - true - if can't increase/reduce value because of insufficient or zero value.
+     */
+    function pushLazyPreview(
+        uint48 node,
+        uint48 begin,
+        uint48 end,
+        uint48 l,
+        uint48 r,
+        uint128 amount,
+        bool isSub
+    ) internal view returns (bool isUpdateInsufficient) {
+        // if reducing and left node is insufficient in funds
+        // of increasing and left node is ZERO run push scenario (left+right) without excluding for [begin, r] leaves
+        if (
+            (isSub && treeNode[node].amount < amount) ||
+            (!isSub && treeNode[node].amount == 0)
+        ) return true;
+
+        if ((node * 2) >= LIQUIDITYNODES) return false; // leaves level reached
+
+        // if node leafs equal to leaf interval then stop
+        if ((begin == l && end == r) || (begin == end)) return false;
+
+        uint48 mid = (begin + end) / 2;
+
+        if (begin <= l && l <= mid) {
+            if (begin <= r && r <= mid) {
+                // [l,r] in [begin,mid] - all leafs in left child
+                if (pushLazyPreview(node * 2, begin, mid, l, r, amount, isSub))
+                    return true;
+            } else {
+                uint256 lAmount = treeNode[node * 2].amount;
+                // get right amount excluding unused leaves
+                uint256 rAmount = treeNode[node * 2 + 1].amount -
+                    getLeavesAmount(node * 2 + 1, mid + 1, end, r + 1, end);
+                uint256 sumAmounts = lAmount + rAmount;
+                if (sumAmounts == 0) return true;
+                uint128 forLeftAmount = uint128(
+                    ((amount * lAmount * DECIMALS) / sumAmounts) / DECIMALS
+                );
+
+                // l in [begin,mid] - part in left child
+                if (
+                    pushLazyPreview(
+                        node * 2,
+                        begin,
+                        mid,
+                        l,
+                        mid,
+                        forLeftAmount,
+                        isSub
+                    )
+                ) return true;
+
+                // r in [mid+1,end] - part in right child
+                if (
+                    pushLazyPreview(
+                        node * 2 + 1,
+                        mid + 1,
+                        end,
+                        mid + 1,
+                        r,
+                        amount - forLeftAmount,
+                        isSub
+                    )
+                ) return true;
```

Comment:
`return pushLazyPreview( node * 2, begin, mid, l, mid, forLeftAmount, isSub) || pushLazyPreview(node * 2 + 1, mid + 1,  end, mid + 1,  r,  amount - forLeftAmount, isSub);`

Reply:
return must be done only for `true` result, means only if interrupt condition found

Reply:
return must be done only for `true` result, means only if interrupt condition found

Reply:
fixed